### ISSUE TITLE
ioctl uses Errno::result incorrectly

### ISSUE THREAD

**Issue by jethrogb** [nix-rust/nix#612]
The ioctl macros automatically call `Errno::result` on the return value of the `ioctl` syscall. This is not correct. As indicated in [the documentation for `Errno::result`](https://docs.rs/nix/0.8.1/nix/enum.Errno.html#method.result), this must only be used if `-1` is returned as an error value. ioctls can return any value directly from the kernel driver and there are no semantics an arbitrary ioctl needs to adhere to.

**Comment by Susurrus** [comment]
Yeah, that does seem problematic. I'd be a fan of transitioning the implemented `ioctl` code to have dedicated function calls per ioctl (like the [ioctl-rs](https://github.com/dcuddeback/ioctl-rs)) crate does it. Then we could be specific about the error cases for each one instead of having to be generic.

**Comment by jethrogb** [comment]
Possibly. It's not possible to support all ioctls with dedicated functions, so a generic framework still needs to be supported.

**Comment by Susurrus** [comment]
Would you be able to write up a PR for this? I'd be happy to review it. I'm very familiar with `ioctl`s and would prefer to have someone more familiar lead the charge here.

**Comment by posborne** [comment]
I landed the initial version of ioctls which had more of a function approach (https://github.com/nix-rust/nix/blob/c4a9598ec16a01e11e66bea1eb34cea9b4646326/src/sys/ioctl.rs) and then it went through some subsequent changes  which I helped facilitate to some degree to move toward the macros.

I was never quite satisfied with where we ended up.  Ioctls are fickle because there is almost nothing in terms of standardization, just conventions (and some of those very OS-specific).  During this process we definitely discussed trying to handle a great number of specific ioctls that could be detected for a given system but in the end we decided that, for nix at least, it only made sense to define a generic framework.

Even if we defined specifics there will always be cases where a generic low-level framework might also be desired (new kernel APIs, custom kernel modules, ...).  The information available would let us define constants at best most likely (so APIs would most likely still need to be unsafe).

> The ioctl macros automatically call Errno::result on the return value of the ioctl syscall. This is not correct.

One could debate this a little and the decision was not without basis.  The Linux man pages for `ioctl` (http://man7.org/linux/man-pages/man2/ioctl.2.html) explicitly states that -1 indicates an error.

>        Usually, on success zero is returned.  A few ioctl() requests use the
>        return value as an output parameter and return a nonnegative value on
>        success.  On error, -1 is returned, and errno is set appropriately.

This is also documented to be the case for FreeBSD: https://www.freebsd.org/cgi/man.cgi?query=ioctl&sektion=2#end as well as Darwin: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man2/ioctl.2.html
